[Clang/AMDGPU] Zero sized arrays not allowed in HIP device code.

VigneshwarJ opened the pull request:
Added diagnosis to throw error when zero sized arrays are used in the HIP device code. SWDEV-449592

VigneshwarJ committed:
[Clang/AMDGPU] Zero sized arrays not allowed in HIP device code.

Added diagnosis to throw error when zero sized arrays are used in
the HIP device code. SWDEV-449592

JonChesterfield reviewed (approved):
Error looks good. Might want to add a case for "dynamic __shared__" to the test file as the syntax is very close to the case being diagnosed - iirc it's things like

```cuda
extern __shared__ float array[];
```

Some existing handling in C like languages conflates `[]` and `[0]`, so might also want `extern __shared__ float array2[0];`

That's probably already covered by existing HIP test cases though, so if everything else is passing I think we're good. Thanks!

tbaederr commented on clang/lib/Sema/SemaType.cpp:
```diff
@@ -2259,6 +2259,17 @@ QualType Sema::BuildArrayType(QualType T, ArraySizeModifier ASM,
+
+        // zero sized static arrays are not allowed in HIP device functions
+        if (LangOpts.HIP && LangOpts.CUDAIsDevice) {
+          auto *FD = dyn_cast_or_null<FunctionDecl>(CurContext);
+          if (FD && (FD->hasAttr<CUDADeviceAttr>() ||
+                     FD->hasAttr<CUDAGlobalAttr>())) {
```
```suggestion
          if (const auto FD = dyn_cast_if_present(...);FD && (FD->hasAttr<CUDADeviceAttr>() ||
                     FD->hasAttr<CUDAGlobalAttr>())) {
```

efriedma-quic commented on clang/lib/Sema/SemaType.cpp:
```diff
@@ -2259,6 +2259,17 @@ QualType Sema::BuildArrayType(QualType T, ArraySizeModifier ASM,
             << 0 << ArraySize->getSourceRange();
+
+        // zero sized static arrays are not allowed in HIP device functions
+        if (LangOpts.HIP && LangOpts.CUDAIsDevice) {
```
I'm not sure you're diagnosing this in the right place; BuildArrayType gets called at the point we parse the type, not the variable, so you can easily avoid the diagnostic using something like the following:

```
typedef float floatarr[0];
__global__ void global_fun() {
    floatarr array;
}
```

VigneshwarJ replied:
Thanks for pointing it out, moved the logic to SemaVarDeclaration, also checking for the typedef and pointers

VigneshwarJ committed:
fix review comments

moved code to SemaVarDecl also check the pointer types to figure out
its within any typedefs or pointers.

efriedma-quic commented on clang/lib/Sema/SemaDecl.cpp:
```diff
@@ -8714,6 +8714,31 @@ void Sema::CheckVariableDeclarationType(VarDecl *NewVD) {
+      };
+      QualType NextTy = NewVD->getType();
+      while (NextTy->isAnyPointerType() || NextTy->isArrayType() ||
+             NextTy->isReferenceType()) {
```
I'm having trouble imagining how a pointer to a zero-size array could cause issues; there isn't any way to tell in IR that pointer points to a zero-length array.

VigneshwarJ committed:
review comment fixes

efriedma-quic commented on clang/lib/Sema/SemaDecl.cpp:
```diff
@@ -8714,6 +8714,21 @@ void Sema::CheckVariableDeclarationType(VarDecl *NewVD) {
+    if (FunctionDecl *FD = getCurFunctionDecl();
+        FD &&
+        (FD->hasAttr<CUDADeviceAttr>() || FD->hasAttr<CUDAGlobalAttr>())) {
+      if (QualType NextTy = NewVD->getType(); NextTy->isArrayType()) {
+        if (const ConstantArrayType *ArrayT =
+                getASTContext().getAsConstantArrayType(NextTy);
```
```suggestion
      if (const ConstantArrayType *ArrayT =
              getASTContext().getAsConstantArrayType(T);
```

efriedma-quic commented on clang/lib/Sema/SemaDecl.cpp:
```diff
@@ -8714,6 +8714,21 @@ void Sema::CheckVariableDeclarationType(VarDecl *NewVD) {
+  if (LangOpts.CUDAIsDevice && LangOpts.HIP) {
+    if (FunctionDecl *FD = getCurFunctionDecl();
+        FD &&
+        (FD->hasAttr<CUDADeviceAttr>() || FD->hasAttr<CUDAGlobalAttr>())) {
```
Is there not some common helper for "will the current context be built for device"?  This seems like it should be common enough to show up in multiple places.

VigneshwarJ replied:
Theres one for CUDA, but I didn't see one for LangOpts.HIP.

efriedma-quic replied:
I'd prefer some reuse, even if you need some additional checks.

VigneshwarJ replied:
The reuse of that DeclAttrsMatchCUDAMode with some rechecks doesn't make sense because that function returns true if decl is not defined. So, inturn I have to check if its inside a functionDecl again else it will wrongly diagnose host global zero array as not permitted.

VigneshwarJ committed:
review comments

VigneshwarJ committed:
Merge branch 'main' into SWDEV_449592

efriedma-quic commented on clang/lib/Sema/SemaDecl.cpp:
```diff
@@ -8732,6 +8732,18 @@ void Sema::CheckVariableDeclarationType(VarDecl *NewVD) {
 
+  // zero sized static arrays are not allowed in HIP device functions
+  if (getLangOpts().HIP &&
+      DeclAttrsMatchCUDAMode(getLangOpts(), getCurFunctionDecl())) {
```
Do you need to check LangOpts.CUDAIsDevice?

VigneshwarJ replied:
Yes and that is being checked in the function DeclAttrsMatchCUDAMode.

efriedma-quic replied:
The check there is `isDeviceSideDecl == LangOpts.CUDAIsDevice`... but you don't want to diagnose this on host, do you?  (The regression test only checks the -fcuda-is-device case.)

VigneshwarJ replied:
Yes, that's right, I dont want to diagnose this on host, but the device side decl are cuda device specific. So I thought there would not be any problem.

VigneshwarJ replied:
else it would be cleaner to go back to how I was checking previously as this is the only helper function I found that is closely related.

efriedma-quic replied:
The code, in its current state, *will* diagnose on host.  (The testcase doesn't show this because it only runs with `-fcuda-is-device`.)

VigneshwarJ committed:
review changes

VigneshwarJ committed:
Merge branch 'SWDEV_449592' of https://github.com/VigneshwarJ/llvm-project into SWDEV_449592

efriedma-quic reviewed (approved):
LGTM
